File: temporal-io/my-python/MY/1_basic/child_workflow.py
"""
Example: Child Workflow
This demonstrates how to use child workflows for modular, reusable workflow components.
Parent workflow delegates subtasks to child workflows.
"""
import asyncio
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from datetime import timedelta
from typing import List

from temporalio import activity, workflow
from temporalio.client import Client
from temporalio.worker import Worker


@dataclass
class OrderItem:
    product: str
    quantity: int
    price: float


@dataclass
class Order:
    order_id: str
    customer: str
    items: List[OrderItem]


# Activities for processing orders


@activity.defn
def validate_inventory(item: OrderItem) -> bool:
    activity.logger.info(f"Validating inventory for {item.product}")
    # Simulate inventory check
    return True


@activity.defn
def calculate_subtotal(item: OrderItem) -> float:
    activity.logger.info(f"Calculating subtotal for {item.product}")
    return item.quantity * item.price


@activity.defn
def process_payment(total_amount: float, customer: str) -> str:
    activity.logger.info(f"Processing payment of ${total_amount} for {customer}")
    # Simulate payment processing
    return f"Payment successful: ${total_amount:.2f}"


@activity.defn
def ship_order(order_id: str, customer: str) -> str:
    activity.logger.info(f"Shipping order {order_id} to {customer}")
    # Simulate shipping
    return f"Order {order_id} shipped to {customer}"


# Child Workflow: Process a single order item
@workflow.defn
class ProcessOrderItemWorkflow:
    @workflow.run
    async def run(self, item: OrderItem) -> float:
        workflow.logger.info(f"Processing item: {item.product}")

        # Validate inventory
        is_available = await workflow.execute_activity(
            validate_inventory,
            item,
            start_to_close_timeout=timedelta(seconds=10),
        )

        if not is_available:
            workflow.logger.warning(f"Item {item.product} not available")
            return 0.0

        # Calculate subtotal
        subtotal = await workflow.execute_activity(
            calculate_subtotal,
            item,
            start_to_close_timeout=timedelta(seconds=10),
        )

        workflow.logger.info(f"Item {item.product} processed: ${subtotal:.2f}")
        return subtotal


# Parent Workflow: Process entire order using child workflows
@workflow.defn
class ProcessOrderWorkflow:
    @workflow.run
    async def run(self, order: Order) -> str:
        workflow.logger.info(f"Processing order {order.order_id} for {order.customer}")

        # Process each item using child workflows
        subtotals: List[float] = []
        for item in order.items:
            workflow.logger.info(f"Starting child workflow for {item.product}")

            # Execute child workflow for each item
            subtotal = await workflow.execute_child_workflow(
                ProcessOrderItemWorkflow.run,
                item,
                id=f"{order.order_id}-item-{item.product}",
            )
            subtotals.append(subtotal)

        # Calculate total
        total_amount = sum(subtotals)
        workflow.logger.info(f"Total amount: ${total_amount:.2f}")

        # Process payment
        payment_result = await workflow.execute_activity(
            process_payment,
            args=[total_amount, order.customer],
            start_to_close_timeout=timedelta(seconds=10),
        )

        # Ship order
        shipping_result = await workflow.execute_activity(
            ship_order,
            args=[order.order_id, order.customer],
            start_to_close_timeout=timedelta(seconds=10),
        )

        result = f"Order {order.order_id} completed:\n"
        result += f"  - {payment_result}\n"
        result += f"  - {shipping_result}\n"
        result += f"  - Items processed: {len(order.items)}"

        workflow.logger.info(f"Order {order.order_id} processing complete")
        return result


async def main():
    # Start client
    client = await Client.connect("localhost:7233")

    # Run a worker for the workflows
    async with Worker(
        client,
        task_queue="1-basic-child-workflow-task-queue",
        workflows=[ProcessOrderWorkflow, ProcessOrderItemWorkflow],
        activities=[
            validate_inventory,
            calculate_subtotal,
            process_payment,
            ship_order,
        ],
        activity_executor=ThreadPoolExecutor(10),
    ):
        # Create a sample order
        order = Order(
            order_id="ORD-12345",
            customer="John Doe",
            items=[
                OrderItem(product="Laptop", quantity=1, price=999.99),
                OrderItem(product="Mouse", quantity=2, price=29.99),
                OrderItem(product="Keyboard", quantity=1, price=79.99),
            ],
        )

        # Execute the parent workflow
        result = await client.execute_workflow(
            ProcessOrderWorkflow.run,
            order,
            id="1-basic-child-workflow-order-12345",
            task_queue="1-basic-child-workflow-task-queue",
        )
        print(f"\n{result}")


if __name__ == "__main__":
    asyncio.run(main())

File: temporal-io/my-typescript/MY/1_basic/child_workflow.ts
/**
 * Example: Child Workflow
 * This demonstrates how to use child workflows for modular, reusable workflow components.
 * Parent workflow delegates subtasks to child workflows.
 *
 * This file contains:
 * - Data types for Order and OrderItem
 * - Activities for order processing
 * - Child workflow for processing individual items
 * - Parent workflow that orchestrates multiple child workflows
 * - Worker and workflow execution logic
 */

// ============================================================================
// DATA TYPES
// ============================================================================

export interface OrderItem {
  product: string;
  quantity: number;
  price: number;
}

export interface Order {
  orderId: string;
  customer: string;
  items: OrderItem[];
}

// ============================================================================
// ACTIVITIES
// ============================================================================
import { log as activityLog } from '@temporalio/activity';

/**
 * Validate inventory for an item
 */
export async function validateInventory(item: OrderItem): Promise<boolean> {
  activityLog.info('Validating inventory for:', { product: item.product });
  // Simulate inventory check
  return true;
}

/**
 * Calculate subtotal for an item
 */
export async function calculateSubtotal(item: OrderItem): Promise<number> {
  activityLog.info('Calculating subtotal for:', { product: item.product });
  return item.quantity * item.price;
}

/**
 * Process payment for total amount
 */
export async function processPayment(totalAmount: number, customer: string): Promise<string> {
  activityLog.info('Processing payment:', { amount: totalAmount, customer });
  // Simulate payment processing
  return `Payment successful: $${totalAmount.toFixed(2)}`;
}

/**
 * Ship order to customer
 */
export async function shipOrder(orderId: string, customer: string): Promise<string> {
  activityLog.info('Shipping order:', { orderId, customer });
  // Simulate shipping
  return `Order ${orderId} shipped to ${customer}`;
}

// ============================================================================
// WORKFLOWS
// ============================================================================
import { proxyActivities, startChild, log as workflowLog } from '@temporalio/workflow';

// Define activity interface
export interface Activities {
  validateInventory(item: OrderItem): Promise<boolean>;
  calculateSubtotal(item: OrderItem): Promise<number>;
  processPayment(totalAmount: number, customer: string): Promise<string>;
  shipOrder(orderId: string, customer: string): Promise<string>;
}

// Proxy activities with timeout configuration
const {
  validateInventory: validateInventoryActivity,
  calculateSubtotal: calculateSubtotalActivity,
  processPayment: processPaymentActivity,
  shipOrder: shipOrderActivity,
} = proxyActivities<Activities>({
  startToCloseTimeout: '10 seconds',
});

/**
 * Child Workflow: Process a single order item
 */
export async function processOrderItemWorkflow(item: OrderItem): Promise<number> {
  workflowLog.info('Processing item:', { product: item.product });

  // Validate inventory
  const isAvailable = await validateInventoryActivity(item);

  if (!isAvailable) {
    workflowLog.warn('Item not available:', { product: item.product });
    return 0.0;
  }

  // Calculate subtotal
  const subtotal = await calculateSubtotalActivity(item);

  workflowLog.info('Item processed:', { product: item.product, subtotal });
  return subtotal;
}

/**
 * Parent Workflow: Process entire order using child workflows
 */
export async function processOrderWorkflow(order: Order): Promise<string> {
  workflowLog.info('Processing order:', { orderId: order.orderId, customer: order.customer });

  // Process each item using child workflows
  const subtotals: number[] = [];

  for (const item of order.items) {
    workflowLog.info('Starting child workflow for:', { product: item.product });

    // Execute child workflow for each item
    const childHandle = await startChild(processOrderItemWorkflow, {
      workflowId: `${order.orderId}-item-${item.product}`,
      args: [item],
    });

    const subtotal = await childHandle.result();
    subtotals.push(subtotal);
  }

  // Calculate total
  const totalAmount = subtotals.reduce((sum, val) => sum + val, 0);
  workflowLog.info('Total amount:', { total: totalAmount });

  // Process payment
  const paymentResult = await processPaymentActivity(totalAmount, order.customer);

  // Ship order
  const shippingResult = await shipOrderActivity(order.orderId, order.customer);

  const result = `Order ${order.orderId} completed:
  - ${paymentResult}
  - ${shippingResult}
  - Items processed: ${order.items.length}`;

  workflowLog.info('Order processing complete:', { orderId: order.orderId });
  return result;
}

// ============================================================================
// MAIN EXECUTION
// ============================================================================
import { Connection, Client } from '@temporalio/client';
import { NativeConnection, Worker } from '@temporalio/worker';

const TEMPORAL_HOST = 'localhost:7233';
const TASK_QUEUE = '1-basic-child-workflow-task-queue';

async function main() {
  try {
    // Connect to Temporal server
    const connection = await Connection.connect({ address: TEMPORAL_HOST });
    const client = new Client({ connection });
    console.log(`Connected to Temporal at ${TEMPORAL_HOST}`);

    // Create worker connection
    const workerConnection = await NativeConnection.connect({
      address: TEMPORAL_HOST,
    });

    // Create worker
    const worker = await Worker.create({
      connection: workerConnection,
      namespace: 'default',
      taskQueue: TASK_QUEUE,
      workflowsPath: new URL(import.meta.url).pathname,
      activities: {
        validateInventory,
        calculateSubtotal,
        processPayment,
        shipOrder,
      },
    });

    console.log(`Worker started on task queue: ${TASK_QUEUE}`);

    // Run worker in background
    const workerPromise = worker.run();

    // Give worker a moment to start
    await new Promise(resolve => setTimeout(resolve, 1000));

    // Create a sample order
    const order: Order = {
      orderId: 'ORD-12345',
      customer: 'John Doe',
      items: [
        { product: 'Laptop', quantity: 1, price: 999.99 },
        { product: 'Mouse', quantity: 2, price: 29.99 },
        { product: 'Keyboard', quantity: 1, price: 79.99 },
      ],
    };

    // Execute the parent workflow
    const workflowId = '1-basic-child-workflow-order-12345';
    console.log(`\nStarting workflow with ID: ${workflowId}`);

    const handle = await client.workflow.start(processOrderWorkflow, {
      args: [order],
      taskQueue: TASK_QUEUE,
      workflowId,
    });

    console.log('Workflow started, waiting for result...');
    const result = await handle.result();

    console.log(`\nWorkflow completed successfully!`);
    console.log(`\n${result}`);

    // Shutdown worker
    worker.shutdown();
    await workerPromise;
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

// Only run main if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}
